_episodes/05b-advanced-containers.md

```diff
@@ -0,0 +1,315 @@
+---
+title: "Creating More Complex Container Images"
+teaching: 30
+exercises: 30
+questions:
+- "How can I make more complex container images? "
+objectives:
+- "Explain how you can include files within Docker images when you build them."
+- "Explain how you can access files on the Docker host from your Docker containers."
+keypoints:
+- You can include files from your Docker host into your Docker images by using the `COPY` instruction in your `Dockerfile`.
+- Docker allows containers to read and write files from the Docker host.
+---
+
+In order to create and use your own containers, you may need more information than
+our previous example. You may want to use files from outside the container, copy
+those files into the container, and just generally learn a little bit about software
+installation. This episode will cover these. Note that the examples will get gradually
+more and more complex - most day-to-day use of containers can be accomplished
+using the first 1-2 sections on this page.
+
+## Using scripts and files from outside the container
+
+In your shell, change to the `sum` folder in the `docker-intro` folder.
+~~~
+$ cd ~/Desktop/docker-intro/sum
+~~~
+{: .language-bash}
+
+This folder has both a `Dockerfile` and a python script called `sum.py`. Let's say
+we wanted to try running the script using our recently created `alpine-python`
+container.
+
+> ## Running containers
+>
+> What command would we use to run python from the `alpine-python` container?
+{: .challenge}
+
+If we try running the container and Python script, what happens?
+~~~
+$ docker run alice/alpine-python python sum.py
+~~~
+{: .language-bash}
+~~~
+python: can't open file 'sum.py': [Errno 2] No such file or directory
+~~~
+{: .output}
+
+> ## No such file or directory
+>
+> What does the error message mean? Why might the Python inside the container
+> not be able to find or open our script?
+>
+{: challenge}
+
+The problem here is that the container and its file system is separate from our
+host computer's file system. When the container runs, it can't see anything outside
+itself, including any of the files on our computer. In order to use Python
+(inside the container) and our script (outside the container, on our computer),
+we need to create a link between the folder on our computer and the container.
+
+This link is called a "mount" and we can create it using an additional
```

Comment:
Maybe we should compare this to "mount"ing or plugging in an external drive?

Reply:
It might also be worth mentioning before this that the mount links the drives both ways. So new files you create in `/temp` will be immediately accessible outside the container and remain after you exit the container.  Not sure if that should go here or later in the lesson somewhere.

Reply:
Added, could use another read through.

Reply:
Looks good to me!